src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Composition;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.ConvertIfToSwitch;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertIfToSwitch
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertIfToSwitchCodeRefactoringProvider)), Shared]
+    internal sealed partial class CSharpConvertIfToSwitchCodeRefactoringProvider : AbstractConvertIfToSwitchCodeRefactoringProvider
+    {
+        protected override IAnalyzer CreateAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+            => new CSharpAnalyzer(syntaxFacts, semanticModel);
+
+        private sealed class CSharpAnalyzer : Analyzer<StatementSyntax, IfStatementSyntax, ExpressionSyntax>
+        {
+            private int _numberOfSubsequentIfStatementsToRemove = 0;
+
+            public CSharpAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+                : base(syntaxFacts, semanticModel)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_if_to_switch;
+
+            protected override IPattern CreatePatternFromExpression(ExpressionSyntax operand)
+            {
+                switch (operand.Kind())
+                {
+                    case SyntaxKind.EqualsExpression:
+                        {
+                            // Look for the form "x == 5" or "5 == x" where "x" is equivalent to the switch expression.
```

Comment:
So one concern i have is if this whole feature is going to be correct in the presence of 'odd' types.  i.e. types that overload operator == but are not int-like types.  If we convert to a switch, won't we then be moving ot .Equals semantics?  If the type has == but not .Equals... we could be changing semantics.

Note: as this is a 'refactoring' vs a fix, that makes things far less bad.  (since a user has to go explicit invoke this, and there is no concept of 'fix all').  I just want to understand if the possibility exists.